craft_application/launchpad/launchpad.py

```diff
@@ -123,27 +138,94 @@ def get_recipe(
         if isinstance(type_, str):
             type_ = models.RecipeType.__members__.get(type_.upper(), type_)
             type_ = models.RecipeType(type_)
+        if owner is None:
+            owner = self.username
 
         if type_ is models.RecipeType.SNAP:
             return models.SnapRecipe.get(self, name, owner)
         if type_ is models.RecipeType.CHARM:
-            if project:
-                return models.CharmRecipe.get(self, name, owner, project)
-            result = None
-            for recipe in models.CharmRecipe.find(self, owner, name=name):
-                if result is not None:
-                    raise ValueError(
-                        f"Multiple charm recipes for {name!r} found with owner {owner!r}",
-                    )
-                result = recipe
-            if result is not None:
-                return result
-            raise ValueError(
-                f"Could not find charm recipe {name!r} with owner {owner!r}",
-            )
+            if not project:
+                raise ValueError("A charm recipe must be associated with a project.")
+            return models.CharmRecipe.get(self, name, owner, project)
 
         raise TypeError(f"Unknown recipe type: {type_}")
 
     def get_project(self, name: str) -> models.Project:
         """Get a project."""
         return models.Project.get(self, name)
+
+    def new_project(
+        self,
+        name: str,
+        *,
+        title: str,
+        display_name: str,
+        summary: str,
+        description: str | None = None,
+    ) -> models.Project:
+        """Create a new project."""
+        return models.Project.new(
+            self, title, name, display_name, summary, description=description
+        )
+
+    @overload
+    def get_repository(self, *, path: str) -> models.GitRepository: ...
```

Comment:
Are these `...` because they you are going to add them later or is it something to do with the overload decorator?

Reply:
It's convention with the overload decorator. Python doesn't have overloads, and redefining a function like this will simply replace the old one. The [`@overload` decorator](https://docs.python.org/3/library/typing.html#typing.overload) is purely for type checkers, and convention is to fill the overloads with an ellipsis. Here I'm adding the overloads so static checkers will provide more useful warnings to developers using `get_repository`.